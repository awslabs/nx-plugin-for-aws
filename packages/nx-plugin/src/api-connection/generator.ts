/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { joinPathFragments, ProjectConfiguration, Tree } from '@nx/devkit';
import { ApiConnectionGeneratorSchema } from './schema';
import trpcReactGenerator from '../trpc/react/generator';
import { hasExportDeclaration } from '../utils/ast';
import { readToml } from '../utils/toml';
import fastApiReactGenerator from '../py/fast-api/react/generator';
import { readProjectConfigurationUnqualified } from '../utils/nx';

/**
 * List of supported source and target project types for api connections
 */
const SUPPORTED_PROJECT_TYPES = [
  'ts#trpc-api',
  'py#fast-api',
  'react',
] as const;

type ProjectType = (typeof SUPPORTED_PROJECT_TYPES)[number];

type Connection = { source: ProjectType; target: ProjectType };

/**
 * Enumerates the supported project connections
 */
const SUPPORTED_CONNECTIONS = [
  { source: 'react', target: 'ts#trpc-api' },
  { source: 'react', target: 'py#fast-api' },
] satisfies Connection[];

type ConnectionKey = (typeof SUPPORTED_CONNECTIONS)[number] extends infer C
  ? C extends Connection
    ? `${C['source']} -> ${C['target']}`
    : never
  : never;

/**
 * Generators for each connection type
 */
const CONNECTION_GENERATORS = {
  'react -> ts#trpc-api': (tree, options) =>
    trpcReactGenerator(tree, {
      frontendProjectName: options.sourceProject,
      backendProjectName: options.targetProject,
    }),
  'react -> py#fast-api': (tree, options) =>
    fastApiReactGenerator(tree, {
      frontendProjectName: options.sourceProject,
      fastApiProjectName: options.targetProject,
    }),
} satisfies Record<
  ConnectionKey,
  (tree: Tree, options: ApiConnectionGeneratorSchema) => Promise<any>
>;

/**
 * Generator for a connection from a project to an API project
 */
export const apiConnectionGenerator = async (
  tree: Tree,
  options: ApiConnectionGeneratorSchema,
) => {
  const sourceType = determineProjectType(tree, options.sourceProject);
  const targetType = determineProjectType(tree, options.targetProject);

  if (!sourceType || !SUPPORTED_PROJECT_TYPES.includes(sourceType)) {
    throw new Error(
      `This generator does not support selected source project ${options.sourceProject}`,
    );
  }
  if (!targetType || !SUPPORTED_PROJECT_TYPES.includes(targetType)) {
    throw new Error(
      `This generator does not support selected target project ${options.targetProject}`,
    );
  }

  const connection = SUPPORTED_CONNECTIONS.find(
    (c) => c.source === sourceType && c.target === targetType,
  );

  if (!connection) {
    throw new Error(
      `This generator does not support a connection from ${options.sourceProject} (${sourceType}) to ${options.targetProject} (${targetType})`,
    );
  }

  return await CONNECTION_GENERATORS[
    `${connection.source} -> ${connection.target}`
  ](tree, options);
};

/**
 * Determine whether the given project is of a known project type
 */
export const determineProjectType = (
  tree: Tree,
  projectName: string,
): ProjectType | undefined => {
  const projectConfiguration = readProjectConfigurationUnqualified(
    tree,
    projectName,
  );

  // NB: if adding new checks, ensure these go from most to least specific
  // eg. react website is more specific than typescript project

  if (isTrpcApi(tree, projectConfiguration)) {
    return 'ts#trpc-api';
  }

  if (isFastApi(tree, projectConfiguration)) {
    return 'py#fast-api';
  }

  if (isReact(tree, projectConfiguration)) {
    return 'react';
  }

  return undefined;
};

const sourceRoot = (projectConfiguration: ProjectConfiguration) =>
  projectConfiguration.sourceRoot ??
  joinPathFragments(projectConfiguration.root, 'src');

const isTrpcApi = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // If the project.json says it's a trpc api, there's no need for introspection
  if ((projectConfiguration.metadata as any)?.apiType === 'trpc') {
    return true;
  }

  // Find the src/index.ts
  const indexTs = tree.read(
    joinPathFragments(sourceRoot(projectConfiguration), 'index.ts'),
    'utf-8',
  );
  if (indexTs === null) {
    return false;
  }

  // If the index file exports an AppRouter, it's a trpc api
  if (hasExportDeclaration(indexTs, 'AppRouter')) {
    return true;
  }

  // If there's a src/router.ts or src/lambdas/router.ts which exports an AppRouter, it's a trpc api
  const trpcRouter =
    tree.read(
      joinPathFragments(sourceRoot(projectConfiguration), 'router.ts'),
      'utf-8',
    ) ??
    tree.read(
      joinPathFragments(
        sourceRoot(projectConfiguration),
        'lambdas',
        'router.ts',
      ),
      'utf-8',
    );

  if (trpcRouter === null) {
    return false;
  }

  if (hasExportDeclaration(trpcRouter, 'AppRouter')) {
    return true;
  }

  return false;
};

const isReact = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // if there's a main.tsx, it's a react app
  return tree.exists(
    joinPathFragments(sourceRoot(projectConfiguration), 'main.tsx'),
  );
};

const isFastApi = (
  tree: Tree,
  projectConfiguration: ProjectConfiguration,
): boolean => {
  // If the project.json says it's a fast api, there's no need for introspection
  if ((projectConfiguration.metadata as any)?.apiType === 'fast-api') {
    return true;
  }

  const pyProjectPath = joinPathFragments(
    projectConfiguration.root,
    'pyproject.toml',
  );
  if (tree.exists(pyProjectPath)) {
    const pyProject = readToml(tree, pyProjectPath);
    if (((pyProject as any)?.project?.dependencies ?? []).includes('fastapi')) {
      return true;
    }
  }

  return false;
};

export default apiConnectionGenerator;
