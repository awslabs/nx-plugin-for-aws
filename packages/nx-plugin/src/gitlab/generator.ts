/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, Tree, joinPathFragments } from '@nx/devkit';
import { GitlabGeneratorSchema } from './schema';

export async function gitlabGenerator(
  tree: Tree,
  options: GitlabGeneratorSchema
) {
  generateFiles(tree, joinPathFragments(__dirname, 'files'), '.', options);
}

export default gitlabGenerator;
