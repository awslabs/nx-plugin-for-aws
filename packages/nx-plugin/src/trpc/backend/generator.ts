/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  getPackageManagerCommand,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  Tree,
  updateJson,
} from '@nx/devkit';
import { TrpcBackendGeneratorSchema } from './schema';
import kebabCase from 'lodash.kebabcase';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
  sharedConstructsGenerator,
} from '../../utils/shared-constructs';
import tsLibGenerator from '../../ts/lib/generator';
import { getNpmScopePrefix, toScopeAlias } from '../../utils/npm-scope';
import { withVersions } from '../../utils/versions';
import { getRelativePathToRoot } from '../../utils/paths';
import { toClassName } from '../../utils/names';
import { addStarExport } from '../../utils/ast';
import { formatFilesInSubtree } from '../../utils/format';
export async function trpcBackendGenerator(
  tree: Tree,
  options: TrpcBackendGeneratorSchema,
) {
  await sharedConstructsGenerator(tree);
  const apiNamespace = getNpmScopePrefix(tree);
  const apiNameKebabCase = kebabCase(options.apiName);
  const apiNameClassName = toClassName(options.apiName);
  const projectRoot = joinPathFragments(
    options.directory ?? '.',
    apiNameKebabCase,
  );
  const relativePathToProjectRoot = `${joinPathFragments(
    getRelativePathToRoot(tree, `${getNpmScopePrefix(tree)}common-constructs`),
    projectRoot,
  )}`;
  const schemaRoot = joinPathFragments(projectRoot, 'schema');
  const backendRoot = joinPathFragments(projectRoot, 'backend');
  const backendName = `${apiNameKebabCase}-backend`;
  const schemaName = `${apiNameKebabCase}-schema`;
  const backendProjectName = `${apiNamespace}${backendName}`;
  const schemaProjectName = `${apiNamespace}${schemaName}`;
  const enhancedOptions = {
    backendProjectName,
    backendProjectAlias: toScopeAlias(backendProjectName),
    schemaProjectName,
    schemaProjectAlias: toScopeAlias(schemaProjectName),
    apiNameKebabCase,
    apiNameClassName,
    relativePathToProjectRoot,
    pkgMgrCmd: getPackageManagerCommand().exec,
    ...options,
  };
  await tsLibGenerator(tree, {
    name: backendName,
    directory: projectRoot,
    subDirectory: 'backend',
  });
  await tsLibGenerator(tree, {
    name: schemaName,
    directory: projectRoot,
    subDirectory: 'schema',
  });
  if (
    !tree.exists(
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'trpc-apis',
        `${apiNameKebabCase}.ts`,
      ),
    )
  ) {
    generateFiles(
      tree,
      joinPathFragments(
        __dirname,
        'files',
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
      ),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'app'),
      enhancedOptions,
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );
    const shouldGenerateCoreTrpcApiConstruct = !tree.exists(
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'core',
        'trpc-api.ts',
      ),
    );
    if (shouldGenerateCoreTrpcApiConstruct) {
      generateFiles(
        tree,
        joinPathFragments(
          __dirname,
          'files',
          SHARED_CONSTRUCTS_DIR,
          'src',
          'core',
        ),
        joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'core'),
        enhancedOptions,
        {
          overwriteStrategy: OverwriteStrategy.KeepExisting,
        },
      );
    }
    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'index.ts',
      ),
      './trpc-apis/index.js',
    );
    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'trpc-apis',
        'index.ts',
      ),
      `./${apiNameKebabCase}.js`,
    );
    if (shouldGenerateCoreTrpcApiConstruct) {
      addStarExport(
        tree,
        joinPathFragments(
          PACKAGES_DIR,
          SHARED_CONSTRUCTS_DIR,
          'src',
          'core',
          'index.ts',
        ),
        './trpc-api.js',
      );
    }
  }

  updateJson(
    tree,
    joinPathFragments(backendRoot, 'project.json'),
    (config: ProjectConfiguration) => {
      config.metadata = {
        apiName: options.apiName,
      } as unknown;
      config.targets = {
        ...config.targets,
        serve: {
          executor: 'nx:run-commands',
          options: {
            commands: ['tsx src/local-server.ts'],
            cwd: backendRoot,
          },
        },
      };
      return config;
    },
  );

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'backend'),
    backendRoot,
    enhancedOptions,
    {
      overwriteStrategy: OverwriteStrategy.Overwrite,
    },
  );
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'schema'),
    schemaRoot,
    enhancedOptions,
    {
      overwriteStrategy: OverwriteStrategy.Overwrite,
    },
  );
  tree.delete(joinPathFragments(backendRoot, 'src', 'lib'));
  tree.delete(joinPathFragments(schemaRoot, 'src', 'lib'));
  addDependenciesToPackageJson(
    tree,
    withVersions([
      'aws-xray-sdk-core',
      'zod',
      '@aws-lambda-powertools/logger',
      '@aws-lambda-powertools/metrics',
      '@aws-lambda-powertools/tracer',
      '@trpc/server',
    ]),
    withVersions(['@types/aws-lambda', 'tsx']),
  );
  tree.delete(joinPathFragments(backendRoot, 'package.json'));
  tree.delete(joinPathFragments(schemaRoot, 'package.json'));
  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
}
export default trpcBackendGenerator;
