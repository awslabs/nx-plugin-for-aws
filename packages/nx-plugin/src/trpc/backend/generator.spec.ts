/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { readProjectConfiguration, Tree } from '@nx/devkit';
import { TRPC_BACKEND_GENERATOR_INFO, tsTrpcApiGenerator } from './generator';
import {
  createTreeUsingTsSolutionSetup,
  snapshotTreeDir,
} from '../../utils/test';
import { expectHasMetricTags } from '../../utils/metrics.spec';

describe('trpc backend generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  it('should generate backend and schema projects', async () => {
    await tsTrpcApiGenerator(tree, {
      name: 'TestApi',
      directory: 'apps',
      computeType: 'ServerlessApiGatewayHttpApi',
    });

    // Verify project structure
    expect(tree.exists('apps/test-api/backend')).toBeTruthy();
    expect(tree.exists('apps/test-api/schema')).toBeTruthy();

    // Verify generated files
    expect(tree.exists('apps/test-api/backend/src/index.ts')).toBeTruthy();
    expect(tree.exists('apps/test-api/backend/src/procedures')).toBeTruthy();
    expect(tree.exists('apps/test-api/schema/src/index.ts')).toBeTruthy();

    // Create snapshots of generated files
    snapshotTreeDir(tree, 'apps/test-api/backend/src');
    snapshotTreeDir(tree, 'apps/test-api/schema/src');
  });

  it('should set up project configuration correctly', async () => {
    await tsTrpcApiGenerator(tree, {
      name: 'TestApi',
      directory: 'apps',
      computeType: 'ServerlessApiGatewayHttpApi',
    });
    const backendProjectConfig = JSON.parse(
      tree.read('apps/test-api/backend/project.json', 'utf-8'),
    );
    // Verify project metadata
    expect(backendProjectConfig.metadata).toEqual({
      apiName: 'TestApi',
      apiType: 'trpc',
    });
  });

  it('should add required dependencies', async () => {
    await tsTrpcApiGenerator(tree, {
      name: 'TestApi',
      directory: 'apps',
      computeType: 'ServerlessApiGatewayHttpApi',
    });
    const packageJson = JSON.parse(tree.read('package.json', 'utf-8'));
    // Verify dependencies were added
    expect(packageJson.dependencies['@trpc/server']).toBeDefined();
    expect(packageJson.dependencies['zod']).toBeDefined();
    expect(packageJson.dependencies['aws-xray-sdk-core']).toBeDefined();
    expect(packageJson.dependencies['aws-cdk-lib']).toBeDefined();
    expect(packageJson.dependencies['constructs']).toBeDefined();
    expect(
      packageJson.dependencies['@aws-lambda-powertools/logger'],
    ).toBeDefined();
    expect(
      packageJson.dependencies['@aws-lambda-powertools/metrics'],
    ).toBeDefined();
    expect(
      packageJson.dependencies['@aws-lambda-powertools/tracer'],
    ).toBeDefined();
    expect(packageJson.devDependencies['@types/aws-lambda']).toBeDefined();
  });

  it('should set up shared constructs for http', async () => {
    await tsTrpcApiGenerator(tree, {
      name: 'TestApi',
      directory: 'apps',
      computeType: 'ServerlessApiGatewayHttpApi',
    });
    // Verify shared constructs setup
    expect(
      tree.exists('packages/common/constructs/src/app/apis/index.ts'),
    ).toBeTruthy();
    expect(
      tree.exists('packages/common/constructs/src/app/apis/test-api.ts'),
    ).toBeTruthy();

    expect(
      tree.read('packages/common/constructs/src/app/apis/index.ts', 'utf-8'),
    ).toContain("export * from './test-api.js'");
    expect(
      tree.read('packages/common/constructs/src/app/index.ts', 'utf-8'),
    ).toContain("export * from './apis/index.js'");
    expect(
      tree.read('packages/common/constructs/src/app/apis/test-api.ts', 'utf-8'),
    ).toMatchSnapshot('test-api.ts');
    expect(
      tree.read('packages/common/constructs/src/core/api/http-api.ts', 'utf-8'),
    ).toMatchSnapshot('http-api.ts');
    expect(
      tree.read('packages/common/constructs/src/core/api/utils.ts', 'utf-8'),
    ).toMatchSnapshot('utils.ts');
    expect(
      tree.read(
        'packages/common/constructs/src/core/api/trpc-utils.ts',
        'utf-8',
      ),
    ).toMatchSnapshot('trpc-utils.ts');

    expect(
      tree.exists('packages/common/constructs/src/core/api/rest-api.ts'),
    ).toBeFalsy();
  });

  it('should set up shared constructs for rest', async () => {
    await tsTrpcApiGenerator(tree, {
      name: 'TestApi',
      directory: 'apps',
      computeType: 'ServerlessApiGatewayRestApi',
    });
    // Verify shared constructs setup
    expect(
      tree.exists('packages/common/constructs/src/app/apis/index.ts'),
    ).toBeTruthy();
    expect(
      tree.exists('packages/common/constructs/src/app/apis/test-api.ts'),
    ).toBeTruthy();

    expect(
      tree.read('packages/common/constructs/src/app/apis/index.ts', 'utf-8'),
    ).toContain("export * from './test-api.js'");
    expect(
      tree.read('packages/common/constructs/src/app/index.ts', 'utf-8'),
    ).toContain("export * from './apis/index.js'");
    expect(
      tree.read('packages/common/constructs/src/app/apis/test-api.ts', 'utf-8'),
    ).toMatchSnapshot('test-api.ts');
    expect(
      tree.read('packages/common/constructs/src/core/api/rest-api.ts', 'utf-8'),
    ).toMatchSnapshot('rest-api.ts');
    expect(
      tree.read('packages/common/constructs/src/core/api/utils.ts', 'utf-8'),
    ).toMatchSnapshot('utils.ts');
    expect(
      tree.read(
        'packages/common/constructs/src/core/api/trpc-utils.ts',
        'utf-8',
      ),
    ).toMatchSnapshot('trpc-utils.ts');

    expect(
      tree.exists('packages/common/constructs/src/core/api/http-api.ts'),
    ).toBeFalsy();
  });

  it('should add a task for starting a local server', async () => {
    await tsTrpcApiGenerator(tree, {
      name: 'TestApi',
      directory: 'apps',
      computeType: 'ServerlessApiGatewayHttpApi',
    });
    const projectConfig = readProjectConfiguration(tree, '@proj/test-api');
    expect(projectConfig.targets).toHaveProperty('serve');
    expect(projectConfig.targets!.serve!.executor).toBe('nx:run-commands');
    expect(projectConfig.targets!.serve!.options!.commands).toEqual([
      'tsx src/local-server.ts',
    ]);
  });

  it('should add generator metric to app.ts', async () => {
    // Call the generator function
    await tsTrpcApiGenerator(tree, {
      name: 'TestApi',
      directory: 'apps',
      computeType: 'ServerlessApiGatewayHttpApi',
    });

    // Verify the metric was added to app.ts
    expectHasMetricTags(tree, TRPC_BACKEND_GENERATOR_INFO.metric);
  });

  it('should include CORS headers in router.ts when using REST API', async () => {
    await tsTrpcApiGenerator(tree, {
      name: 'TestApi',
      directory: 'apps',
      computeType: 'ServerlessApiGatewayRestApi',
    });

    // Read the generated router.ts file
    const routerTsContent = tree.read(
      'apps/test-api/backend/src/router.ts',
      'utf-8',
    );

    // Verify CORS headers are included in responseMeta
    expect(routerTsContent).toContain('responseMeta: () => ({');
    expect(routerTsContent).toContain("'Access-Control-Allow-Origin': '*'");
    expect(routerTsContent).toContain("'Access-Control-Allow-Methods': '*'");
  });
});
