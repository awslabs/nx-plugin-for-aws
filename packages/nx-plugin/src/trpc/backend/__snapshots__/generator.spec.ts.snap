// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/client/index.ts 1`] = `
"import {
  createTRPCClient,
  httpBatchLink,
  httpLink,
  HTTPBatchLinkOptions,
  HTTPLinkOptions,
  splitLink,
} from '@trpc/client';

import { AppRouter } from '../router.js';
import { sigv4Fetch } from './sigv4.js';

export interface TestApiClientConfig {
  readonly url: string;
}

export const createTestApiClient = (config: TestApiClientConfig) => {
  const linkOptions: HTTPLinkOptions<any> & HTTPBatchLinkOptions<any> = {
    url: config.url,
    fetch: sigv4Fetch,
  };
  return createTRPCClient<AppRouter>({
    links: [
      splitLink({
        condition(op) {
          return op.context.skipBatch === true;
        },
        true: httpLink(linkOptions),
        false: httpBatchLink(linkOptions),
      }),
    ],
  });
};
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/client/sigv4.ts 1`] = `
"import { AwsClient } from 'aws4fetch';
import { fromNodeProviderChain } from '@aws-sdk/credential-providers';

const credentialProvider = fromNodeProviderChain();

export const sigv4Fetch = (async (...args) => {
  const client = new AwsClient(await credentialProvider());
  return client.fetch(...args);
}) satisfies AwsClient['fetch'];
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/index.ts 1`] = `
"export type { AppRouter } from './router.js';
export type { Context } from './init.js';
export * from './client/index.js';
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/init.ts 1`] = `
"import { initTRPC } from '@trpc/server';
import {
  createErrorPlugin,
  createLoggerPlugin,
  createMetricsPlugin,
  createTracerPlugin,
  IMiddlewareContext,
} from './middleware/index.js';

process.env.POWERTOOLS_SERVICE_NAME = 'TestApi';
process.env.POWERTOOLS_METRICS_NAMESPACE = 'TestApi';

export type Context = IMiddlewareContext;

export const t = initTRPC.context<Context>().create();

export const publicProcedure = t.procedure
  .unstable_concat(createLoggerPlugin())
  .unstable_concat(createTracerPlugin())
  .unstable_concat(createMetricsPlugin())
  .unstable_concat(createErrorPlugin());
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/local-server.ts 1`] = `
"import { createHTTPServer } from '@trpc/server/adapters/standalone';
import { appRouter } from './router.js';

const PORT = 2022;

createHTTPServer({
  router: appRouter,
  createContext() {
    return {
      event: {} as any,
      context: {} as any,
      info: {} as any,
    };
  },
}).listen(PORT);

console.log(\`Local TRPC server listening on port \${PORT}\`);
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/middleware/error.ts 1`] = `
"import { initTRPC, TRPCError } from '@trpc/server';

export const createErrorPlugin = () => {
  const t = initTRPC.context().create();
  return t.procedure.use(async (opts) => {
    const response = await opts.next({
      ctx: {
        ...opts.ctx,
      },
    });

    if (!response.ok) {
      if (response.error instanceof TRPCError) {
        throw response.error;
      }

      throw new TRPCError({
        code: 'INTERNAL_SERVER_ERROR',
        message: 'An unexpected error occurred, please try again later.',
        cause: response.error,
      });
    }

    return response;
  });
};
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/middleware/index.ts 1`] = `
"import { CreateAWSLambdaContextOptions } from '@trpc/server/adapters/aws-lambda';
import type { APIGatewayProxyEventV2WithIAMAuthorizer } from 'aws-lambda';
import { ILoggerContext } from './logger.js';
import { IMetricsContext } from './metrics.js';
import { ITracerContext } from './tracer.js';

export * from './logger.js';
export * from './metrics.js';
export * from './tracer.js';
export * from './error.js';

export type IMiddlewareContext =
  CreateAWSLambdaContextOptions<APIGatewayProxyEventV2WithIAMAuthorizer> &
    ILoggerContext &
    IMetricsContext &
    ITracerContext;
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/middleware/logger.ts 1`] = `
"import { initTRPC } from '@trpc/server';
import { Logger } from '@aws-lambda-powertools/logger';

export interface ILoggerContext {
  logger?: Logger;
}

export const createLoggerPlugin = () => {
  const t = initTRPC.context<ILoggerContext>().create();
  return t.procedure.use(async (opts) => {
    const logger = new Logger();

    const response = await opts.next({
      ctx: {
        ...opts.ctx,
        logger,
      },
    });

    if (!response.ok) {
      logger.error('Unexpected error occured', response.error.cause ?? '');
    }

    return response;
  });
};
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/middleware/metrics.ts 1`] = `
"import { initTRPC } from '@trpc/server';
import { Metrics, MetricUnit } from '@aws-lambda-powertools/metrics';

export interface IMetricsContext {
  metrics?: Metrics;
}

export const createMetricsPlugin = () => {
  const t = initTRPC.context<IMetricsContext>().create();

  return t.procedure.use(async (opts) => {
    const metrics = new Metrics();
    metrics.addDimensions({
      procedure: opts.path,
      type: opts.type,
    });
    metrics.captureColdStartMetric();
    metrics.addMetric('RequestCount', MetricUnit.Count, 1);

    try {
      const response = await opts.next({
        ctx: {
          ...opts.ctx,
          metrics,
        },
      });

      if (!response.ok) {
        metrics.addMetric('Failure', MetricUnit.Count, 1);
      } else {
        metrics.addMetric('Success', MetricUnit.Count, 1);
      }

      return response;
    } finally {
      metrics.publishStoredMetrics();
    }
  });
};
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/middleware/tracer.ts 1`] = `
"import { initTRPC } from '@trpc/server';
import { Tracer } from '@aws-lambda-powertools/tracer';
import { Subsegment } from 'aws-xray-sdk-core';

export interface ITracerContext {
  tracer?: Tracer;
}

export const createTracerPlugin = () => {
  const t = initTRPC.context<ITracerContext>().create();

  return t.procedure.use(async (opts) => {
    const tracer = new Tracer();
    const segment = tracer.getSegment();
    let handlerSegment: Subsegment | undefined;

    if (segment) {
      handlerSegment = segment.addNewSubsegment(\`## \${opts.path}\`);
      tracer.setSegment(handlerSegment);
    }

    tracer.annotateColdStart();
    tracer.addServiceNameAnnotation();

    try {
      const response = await opts.next({
        ctx: {
          ...opts.ctx,
          tracer,
        },
      });

      if (!response.ok && response.error.cause instanceof Error) {
        tracer.addErrorAsMetadata(response.error.cause);
      }
      return response;
    } finally {
      if (segment && handlerSegment) {
        handlerSegment.close();
        tracer.setSegment(segment);
      }
    }
  });
};
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/procedures/echo.ts 1`] = `
"import { publicProcedure } from '../init.js';
import { EchoInputSchema, EchoOutputSchema } from ':proj/test-api-schema';

export const echo = publicProcedure
  .input(EchoInputSchema)
  .output(EchoOutputSchema)
  .query((opts) => ({ result: opts.input.message }));
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/backend/src/router.ts 1`] = `
"import {
  awsLambdaRequestHandler,
  CreateAWSLambdaContextOptions,
} from '@trpc/server/adapters/aws-lambda';
import { echo } from './procedures/echo.js';
import { t } from './init.js';
import { APIGatewayProxyEventV2WithIAMAuthorizer } from 'aws-lambda';

export const router = t.router;

export const appRouter = router({
  echo,
});

export const handler = awsLambdaRequestHandler({
  router: appRouter,
  createContext: (
    ctx: CreateAWSLambdaContextOptions<APIGatewayProxyEventV2WithIAMAuthorizer>,
  ) => ctx,
});

export type AppRouter = typeof appRouter;
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/schema/src/index.ts 1`] = `
"export * from './procedures/echo.js';
"
`;

exports[`trpc backend generator > should generate backend and schema projects > apps/test-api/schema/src/procedures/echo.ts 1`] = `
"import { z } from 'zod';

export const EchoInputSchema = z.object({
  message: z.string(),
});

export type IEchoInput = z.TypeOf<typeof EchoInputSchema>;

export const EchoOutputSchema = z.object({
  result: z.string(),
});

export type IEchoOutput = z.TypeOf<typeof EchoOutputSchema>;
"
`;

exports[`trpc backend generator > should set up shared constructs > http-api.ts 1`] = `
"import { Construct } from 'constructs';
import { CfnOutput, Duration } from 'aws-cdk-lib';
import {
  CorsHttpMethod,
  HttpApi as _HttpApi,
  HttpMethod,
  IHttpRouteAuthorizer,
} from 'aws-cdk-lib/aws-apigatewayv2';
import { HttpLambdaIntegration } from 'aws-cdk-lib/aws-apigatewayv2-integrations';
import { Code, Function, Runtime, Tracing } from 'aws-cdk-lib/aws-lambda';
import { Grant, IGrantable } from 'aws-cdk-lib/aws-iam';
import { RuntimeConfig } from './runtime-config.js';

export interface HttpApiProps {
  readonly apiName: string;
  readonly handler: string;
  readonly handlerFilePath: string;
  readonly runtime: Runtime;
  readonly defaultAuthorizer: IHttpRouteAuthorizer;
  readonly allowedOrigins?: string[];
}

export class HttpApi extends Construct {
  public readonly api: _HttpApi;
  public readonly routerFunction: Function;

  constructor(scope: Construct, id: string, props: HttpApiProps) {
    super(scope, id);

    this.routerFunction = new Function(this, \`\${id}Handler\`, {
      timeout: Duration.seconds(30),
      runtime: props.runtime,
      handler: props.handler,
      code: Code.fromAsset(props.handlerFilePath),
      tracing: Tracing.ACTIVE,
      environment: {
        AWS_CONNECTION_REUSE_ENABLED: '1',
      },
    });

    this.api = new _HttpApi(this, id, {
      corsPreflight: {
        allowOrigins: props.allowedOrigins ?? ['*'],
        allowMethods: [CorsHttpMethod.ANY],
        allowHeaders: [
          'authorization',
          'content-type',
          'x-amz-content-sha256',
          'x-amz-date',
          'x-amz-security-token',
        ],
      },
      defaultAuthorizer: props.defaultAuthorizer,
    });

    this.api.addRoutes({
      path: '/{proxy+}',
      methods: [
        HttpMethod.GET,
        HttpMethod.DELETE,
        HttpMethod.POST,
        HttpMethod.PUT,
        HttpMethod.PATCH,
        HttpMethod.HEAD,
      ],
      integration: new HttpLambdaIntegration(
        'RouterIntegration',
        this.routerFunction,
      ),
    });

    new CfnOutput(this, \`\${props.apiName}Url\`, { value: this.api.url! });

    RuntimeConfig.ensure(this).config.httpApis = {
      ...RuntimeConfig.ensure(this).config.httpApis!,
      [props.apiName]: this.api.url!,
    };
  }

  public grantInvokeAccess(grantee: IGrantable) {
    Grant.addToPrincipal({
      grantee,
      actions: ['execute-api:Invoke'],
      resourceArns: [this.api.arnForExecuteApi('*', '/*', '*')],
    });
  }
}
"
`;

exports[`trpc backend generator > should set up shared constructs > index.ts 1`] = `
"export * from './test-api.js';
"
`;

exports[`trpc backend generator > should set up shared constructs > test-api.ts 1`] = `
"import { Construct } from 'constructs';
import * as url from 'url';
import { HttpApi } from '../../core/http-api.js';
import { HttpIamAuthorizer } from 'aws-cdk-lib/aws-apigatewayv2-authorizers';
import { Runtime } from 'aws-cdk-lib/aws-lambda';

export class TestApi extends HttpApi {
  constructor(scope: Construct, id: string) {
    super(scope, id, {
      defaultAuthorizer: new HttpIamAuthorizer(),
      apiName: 'TestApi',
      runtime: Runtime.NODEJS_LATEST,
      handler: 'index.handler',
      handlerFilePath: url.fileURLToPath(
        new URL(
          '../../../../../../dist/apps/test-api/backend/bundle',
          import.meta.url,
        ),
      ),
    });
  }
}
"
`;
